src/promql/src/extension_plan/instant_manipulator.rs

```diff
@@ -0,0 +1,721 @@
+    };
+    use datafusion::from_slice::FromSlice;
+    use datafusion::physical_plan::memory::MemoryExec;
+    use datafusion::prelude::SessionContext;
+    use datatypes::arrow::array::TimestampMillisecondArray;
+    use datatypes::arrow_array::StringArray;
+
+    use super::*;
+
+    const TIME_INDEX_COLUMN: &str = "timestamp";
+
+    fn prepare_test_data() -> MemoryExec {
+        let schema = Arc::new(Schema::new(vec![
+            Field::new(TIME_INDEX_COLUMN, TimestampMillisecondType::DATA_TYPE, true),
+            Field::new("value", DataType::Float64, true),
+            Field::new("path", DataType::Utf8, true),
+        ]));
+        let timestamp_column = Arc::new(TimestampMillisecondArray::from_slice([
+            0, 30_000, 60_000, 90_000, 120_000, // every 30s
+            180_000, 240_000, // every 60s
+            241_000, 271_000, 291_000, // others
+        ])) as _;
+        let value_column = Arc::new(Float64Array::from_slice([1.0; 10])) as _;
+        let path_column = Arc::new(StringArray::from_slice(["foo"; 10])) as _;
+        let data = RecordBatch::try_new(
+            schema.clone(),
+            vec![timestamp_column, value_column, path_column],
+        )
+        .unwrap();
+
+        MemoryExec::try_new(&[vec![data]], schema, None).unwrap()
+    }
+
+    async fn do_normalize_test(
+        start: Millisecond,
+        end: Millisecond,
+        lookback_delta: Millisecond,
+        interval: Millisecond,
+        expected: String,
+    ) {
+        let memory_exec = Arc::new(prepare_test_data());
+        let normalize_exec = Arc::new(InstantManipulatorExec {
+            start,
+            end,
+            lookback_delta,
+            interval,
+            time_index_column: TIME_INDEX_COLUMN.to_string(),
+            input: memory_exec,
+            metric: ExecutionPlanMetricsSet::new(),
+        });
+        let session_context = SessionContext::default();
+        let result = datafusion::physical_plan::collect(normalize_exec, session_context.task_ctx())
+            .await
+            .unwrap();
+        let result_literal = datatypes::arrow::util::pretty::pretty_format_batches(&result)
+            .unwrap()
+            .to_string();
+
+        assert_eq!(result_literal, expected);
+    }
+
+    #[tokio::test]
+    async fn lookback_10s_interval_30s() {
+        let expected = String::from(
+            "+---------------------+-------+------+\
+            \n| timestamp           | value | path |\
+            \n+---------------------+-------+------+\
+            \n| 1970-01-01T00:00:00 | 1     | foo  |\
+            \n| 1970-01-01T00:00:30 | 1     | foo  |\
+            \n| 1970-01-01T00:01:00 | 1     | foo  |\
+            \n| 1970-01-01T00:01:30 | 1     | foo  |\
+            \n| 1970-01-01T00:02:00 | 1     | foo  |\
+            \n| 1970-01-01T00:02:30 |       |      |\
+            \n| 1970-01-01T00:03:00 | 1     | foo  |\
```

Comment:
Is it expected? Should the output contain timestamps like this?
```suggestion
            \n| 1970-01-01T00:02:00 | 1     | foo  |\
            \n| 1970-01-01T00:03:00 | 1     | foo  |\
```

Reply:
Yes it's intended:
>If data is absent in some timestamp, all columns except the time index will left blank.

Prometheus won't generate eval result if there is no timestamp. But if we omit those rows, they will also disapper in all the following stages. I'm not sure if this will cause problems so I keep those empty rows for now. It's relatively easy to change this behavior, it won't be too late to change if we find out we have to omit those rows afterwards.